x/staking/keeper/delegation.go

```diff
@@ -680,8 +680,11 @@ func (k Keeper) Unbond(
 		validator = k.mustGetValidator(ctx, validator.GetOperator())
 	}
 
-	// remove the delegation
-	if delegation.Shares.IsZero() {
+	// Remove the delegation if the resulting shares yield a truncated zero amount
+	// of tokens in the delegation. Note, in this this case, the shares themselves
+	// may not be zero, but rather a small fractional amount. Otherwise, we update
+	// the delegation object.
+	if validator.TokensFromShares(delegation.Shares).TruncateInt().IsZero() || delegation.Shares.IsZero() {
```

Comment:
@alexanderbez do you think this could break delegator shares invariant, causing https://github.com/cosmos/cosmos-sdk/issues/10750 ?

@aleem1314 did some debugging on this, here's some logs:
```
++++++++++++++++++++++++++++++++++++
TOKENS-FROM-SHARES =  0.937500000000000000  validator.TokensFromShares(delegation.Shares).TruncateInt() =  0
DELEGATIONS SHARES =  0.937500000000000000
VALIDATOR =  cosmosvaloper14rmvn8ymw7mdjwyjhmkzestgkcdskvkx6ke9e5
DELEGATOR =  cosmos15rxvlc2xaf6sjq5fjq45gfapdjsjr97jy6wmjc
BLOCKHEIGHT =  32
BLOCKTIME =  2103-10-01 17:18:06 +0000 UTC
++++++++++++++++++++++++++++++++++++
```

just reverting this line and simulations pass.

Reply:
Hmm could be! But what invariant is broken exactly? The delegation should be deleted here. Is there some invariant that expects this not to be deleted?

Reply:
The `DelegatorSharesInvariant` one. Maybe the invariant should be updated to take the `TruncateInt` into account?

Reply:
Yes, let's keep the discussion in the new issue created 👍